src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/TaskCache.cs

```diff
@@ -40,5 +40,14 @@ private static Task<int>[] CreateInt32Tasks()
 
             return tasks;
         }
+
+        /// <summary>
+        /// A cached task for default(TResult).
+        /// </summary>
+        internal static class CachedDefaultValueTask<TResult>
```

Comment:
Does this introduce any perf issues, in particular for reference type TResults, where it's now slower due to one generic context calling into a different generic context?

Reply:
@stephentoub There is no such situation in the existing code, as far as I can tell.

The old `Task<TResult>.s_DefaultResultTask` was only used in 3 places:
1. in Stream.cs we had `Task<int>.s_DefaultResultTask` which is a concerete instantiation
2. in Task.cs we had two uses in `Task.FromResult<TResult>(TResult result)`:
  * `if (result is null) return Task<TResult>.s_DefaultResultTask` was already a generic context calling into a different generic context.  Same as the new code `TaskCache.CachedDefaultTask<TResult>.s_DefaultResultTask`
  * `if (/* result is the same size as one of several primitive value types and the bitpattern matches the default value of that primitive value type */) return Task<TResult>.s_DefaultResultTask;` - again this is already one generic context calling into another generic context, although they're not reference types.

I'll try a microbenchmark on `Task.FromResult<T>(default(T))` for some different types.

Reply:
Added a microbenchmark:

```csharp
[GenericTypeArguments(typeof(byte))]
[GenericTypeArguments(typeof(int))]
[GenericTypeArguments(typeof(string))]
[GenericTypeArguments(typeof(Type))]
[BenchmarkCategory(Categories.Libraries)]
public class Perf_FromResult<T>
{
    [Benchmark]
    public void FromResultDefault()
    {
        Task<T> d = Task.FromResult<T>(default(T));
    }
}
```

(below "runtime-bugs" is baseline, and "runtime" is the PR)

Results for `string`:

| Method            | Job        | Toolchain                                                                                                | Mean     | Error     | StdDev    | Median   | Min      | Max      | Ratio | MannWhitney(5%) | RatioSD | Allocated | Alloc Ratio |
|------------------ |----------- |--------------------------------------------------------------------------------------------------------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|---------------- |--------:|----------:|------------:|
| FromResultDefault | Job-KNJPPY | /runtime-bugs/artifacts/bin/testhost/net9.0-osx-Release-arm64/shared/Microsoft.NETCore.App/9.0.0/corerun | 6.648 ns | 0.1802 ns | 0.2075 ns | 6.568 ns | 6.466 ns | 7.146 ns |  1.00 | Base            |    0.00 |         - |          NA |
| FromResultDefault | Job-SJEICH | /runtime/artifacts/bin/testhost/net9.0-osx-Release-arm64/shared/Microsoft.NETCore.App/9.0.0/corerun      | 6.358 ns | 0.0123 ns | 0.0109 ns | 6.356 ns | 6.344 ns | 6.379 ns |  0.96 | Same            |    0.03 |         - |          NA |

results for `System.Type`:

| Method            | Job        | Toolchain                                                                                                | Mean     | Error     | StdDev    | Median   | Min      | Max      | Ratio | MannWhitney(5%) | Allocated | Alloc Ratio |
|------------------ |----------- |--------------------------------------------------------------------------------------------------------- |---------:|----------:|----------:|---------:|---------:|---------:|------:|---------------- |----------:|------------:|
| FromResultDefault | Job-KNJPPY | /runtime-bugs/artifacts/bin/testhost/net9.0-osx-Release-arm64/shared/Microsoft.NETCore.App/9.0.0/corerun | 6.391 ns | 0.0072 ns | 0.0067 ns | 6.390 ns | 6.384 ns | 6.405 ns |  1.00 | Base            |         - |          NA |
| FromResultDefault | Job-SJEICH | /runtime/artifacts/bin/testhost/net9.0-osx-Release-arm64/shared/Microsoft.NETCore.App/9.0.0/corerun      | 6.368 ns | 0.0136 ns | 0.0113 ns | 6.365 ns | 6.354 ns | 6.398 ns |  1.00 | Same            |         - |          NA |